k2/csrc/determinize.cc

```diff
@@ -7,44 +7,136 @@
+
+  std::shared_ptr<LogSumTracebackState> prev_state;
+};
+
+struct LogSumTracebackState {
+  // LogSumTracebackState can be thought of as as a weighted set of paths from the
+  // start state to a particular state.  (It will be limited to the subset of
+  // paths that have a specific symbol sequence).
+
+  // `prev_elements` is, conceptually, a list of pairs (incoming arc-index,
+  // traceback link); we will keep it free of duplicates of the same incoming
+  // arc.
+  vector<LogSumTracebackLink> prev_elements;
+
+
+  int32_t arc_index;  // Index of most recent arc in path from start-state to
+                      // the dest-state, or -1 if the path is empty (only
+                      // possible if this element belongs to the start-state).
+
+  int32_t symbol;     // Symbol on the arc numbered `arc_index` of the input FSA
+                      // (copied here for convenience), or 0 if arc_index == -1.
+
+  MaxTracebackState(std::shared_ptr<MaxTracebackState> prev,
+              int32_t arc_index, int32_t symbol):
+      prev(prev), arc_index(arc_index), symbol(symbol) { }
+
+};
+
+
+struct DetStateElement {
+
+  double weight;      // Weight from reference state to this state, along
+                      // the path taken by following the 'prev' links
+                      // (the path would have `seq_len` arcs in it).
+                      // Note: by "this state" we mean the destination-state of
+                      // the arc at `arc_index`.
+                      // Interpret this with caution, because the
+                      // base state, and the length of the sequence arcs from the
+                      // base state to here, are known only in the DetState
+                      // that owns this DetStateElement.
+
+  std::shared_ptr<PathElement> path;
+                      // The path from the start state to here (actually we will
+                      // only follow back `seq_len` links.  Will be nullptr if
+                      // seq_len == 0 and this belongs to the initial determinized
+                      // state.
+
+  DetStateElement &&Advance(float arc_weight, int32_t arc_index, int32_t arc_symbol) {
+    return DetStateElement(weight + arc_weight,
+                           std::make_shared<PathElement>(path, arc_index, arc_symbol));
   }
+
+  DetStateElement(double weight, std::shared_ptr<PathElement> &&path):
+      weight(weight), path(path) { }
+
+};
+
+class DetState;
+
+
+struct DetStateCompare {
+  // Comparator for priority queue.  Less-than operator that compares
+  // forward_backward_weight for best-first processing.
+  bool operator()(const shared_ptr<DetState> &a,
+                  const shared_ptr<DetState> &b);
 };
 
+
+
+class Determinizer {
+ public:
+ private:
+
+  using DetStatePriorityQueue = priority_queue<shared_ptr<DetState>,
+                                               vector<shared_ptr<DetState> >,
```

Comment:
Aha, I noticed that you are used to add a blank between `> >` when using STL, actually this is not required anymore in modern C++ compiler (it has been a bug).

Reply:
OK